lib/Dialect/FIRRTL/Transforms/SimplifyAggregates.cpp

```diff
@@ -244,12 +247,44 @@ bool MergeConnection::peelConnect(StrictConnectOp connect) {
   return true;
 }
 
-bool MergeConnection::run() {
+// Return true if multibit mux was folded into subaccess op.
+// If multbit mux inputs are `a[n-1], a[n-2], ..., a[1], a[0]`, we can fold
+// multibit into subaccess op.
+bool SimplifyAggregate::simplifyMultibitMux(MultibitMuxOp multibitMux) {
+
+  // Sample the last element of multibit mux inputs.
+  auto subindex = multibitMux.getInputs().back().getDefiningOp<SubindexOp>();
+  if (subindex.getIndex() != 0)
+    return false;
+  for (auto [idx, value] :
+       llvm::enumerate(multibitMux.getInputs().drop_back())) {
+    auto sub = value.getDefiningOp<SubindexOp>();
+    if (!sub || sub.getIndex() + idx + 1 != multibitMux.getInputs().size())
+      return false;
+  }
+
+  builder->setInsertionPointAfter(multibitMux);
+  builder->setLoc(multibitMux.getLoc());
+  Value subaccess =
+      builder->create<SubaccessOp>(subindex.getInput(), multibitMux.getIndex());
+  multibitMux.replaceAllUsesWith(subaccess);
+  return true;
+}
+
+bool SimplifyAggregate::run() {
   ImplicitLocOpBuilder theBuilder(moduleOp.getLoc(), moduleOp.getContext());
   builder = &theBuilder;
   auto *body = moduleOp.getBodyBlock();
-  // Merge connections by forward iterations.
   for (auto it = body->begin(), e = body->end(); it != e;) {
```

Comment:
This loop probably should be a preincrement safe wrapper loop.

Reply:
No, we cannot do preincrement since new operations might be created within `peelConnect` and we have to visit these operations as well.